auth/design.md

```diff
@@ -0,0 +1,69 @@
+The v3 protocol uses gRPC as its transport instead of RESTful interface like v2. Therefore we have a chance of improvements (e.g. connection based auth instead of request based auth). In addition, there is an unclear point in the semantics of the v2 auth (described later). In v3, it should be cleaned and a well defined guidance should be provided for users and developers.
+
+### Functionality requirements
+1. per connection authentication, not per request
+   1. User ID + password based authentication implemented on gRPC API
+   1. authentication must be refreshed after auth information update
+1. its functionality should be as simple and useful as v2
+   1. v3 provides flat key space, unlike the directory structure of v2. Permission check will be provided as simple prefix matching.
+1. it should avoid the consistency problem of v2, described later
+
+### Main required changes
+1. a client needs to create a dedicated connection just for authentication before sending actual requests
+1. add permission information (user ID and authorized revision) to the Raft commands (`etcdserverpb.InternalRaftRequest`)
+1. every request requires permission check in the state machine layer, not in the API layer
+
+### Consistency of the metadata for auth
+The metadata for auth should also be stored and managed in the storage controlled by etcd's Raft protocol like other data stored in etcd. It is required for not sacrificing availability and consistency of the entire etcd cluster. If reading or writing the metadata (e.g. permission information) needs an agreement of every node (more than quorum), single node failure can stop the entire cluster. Because permission checking during ordinal reading/writing requests cannot be completed if the cluster lacks a single node, even if it has living quorum nodes. It degrades availability of the cluster.
+
+The authentication mechanism in etcd v2 protocol has a tricky part because of the metadata consistency described in the above: permission check is processed in each node that receives HTTP requests from clients (etcdserver/etcdhttp/client.go), even the nodes are followers. Therefore, the check can be performed based on stale metadata.
+
+The tricky part is that auth configuration cannot be reflected as soon as operators execute etcdctl. Therefore there is no way to know how long the stale metadata is living. Practically, the configuration change is reflected immediately after the command execution. However, in some cases of heavy load, the inconsistent state can be enlarged and it might result counter-intuitive situation for users and developers. It requires a workaround like this: https://github.com/coreos/etcd/pull/4317#issuecomment-179037582
+
+### How the inconsistent metadata relates with linearizable read and write
+The inconsistent state should be cared especially for writes. Even an operator disables write for some users, they have a chance of write because of the stale permission information.

```

Comment:
Inconsistent authentication state is perhaps most serious for writes. Even if an operator disables write on a user, if the write is only ordered with respect to the key value store but not the authentication system, it's possible the write will complete successfully. Without ordering on both the auth store and the key-value store, the system will be susceptible to stale permission attacks.